File: src/com/excelsior/nothing/Kernel.java
/*
 * Copyright (c) 2012, Nikita Lipsky, Excelsior LLC.
 *
 *  The Nothing System is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  The Nothing System is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
 *
*/

package com.excelsior.nothing;

import com.excelsior.common.QuotedStringTokenizer;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;

/**
 * @author kit
 * Date: 05.10.12
 */
public class Kernel {

    private static HashMap<String, Object> registry = new HashMap<String, Object>();

    private static HashMap<String, Boolean> cache = new HashMap<String, Boolean>();

    public static boolean isCommand(String line) {
        QuotedStringTokenizer st = new QuotedStringTokenizer(line);
        String cmd = st.nextToken();
        if (cmd == null) {
            return false;
        }
        if (cache.containsKey(cmd)) return cache.get(cmd);
        MethodHandle m = MethodHandle.getMethodHandle(cmd);
        if (m != null)
            cache.put(cmd, m != null);
        return m != null;
    }

    public static void executeCommand(String line) {
        cache.clear();
        QuotedStringTokenizer st = new QuotedStringTokenizer(line);
        String cmd = st.nextToken();
        if (cmd == null) {
            return;
        }
        MethodHandle m = MethodHandle.getMethodHandle(cmd);
        if (m == null) return;

        String firstArg = st.nextToken();
        ArrayList<String> args = new ArrayList<String>();

        if (firstArg != null) {
            if (firstArg.equals("^") && (Main.curSelection != null) && (Main.curSelection.getSelectedText() != null)) {
                st = new QuotedStringTokenizer(Main.curSelection.getSelectedText());
            } else {
                args.add(firstArg);
            }

            while (true) {
                String nextToken = st.nextToken();
                if (nextToken == null) break;
                args.add(nextToken);
            }
        }

        try {
            m.invoke(args.toArray(new String[args.size()]));
        } catch (Exception e) {
//            e.printStackTrace();
        }
    }

    public static void addToRegistry(String key, Object o) {
        registry.put(key, o);
    }

    public static Object getFromRegistry(String key) {
        return registry.get(key);
    }

    private static URL getBaseUrlFrom(URL url)
    {
        try {
            return new URL(url.getProtocol() + "://" + url.getHost());
        } catch (MalformedURLException e) {
            assert false: "Bad base URL built";
            return null;
        }
    }

    public static InputStream getInputStream(String file) throws IOException {
        try {
            URL url = new URL(file);
            MethodHandle.addBaseURL(getBaseUrlFrom(url));
            return url.openStream();
        } catch (MalformedURLException e) {
            return new FileInputStream(file);
        }
    }
}

File: src/main/kotlin/com/excelsior/nothing/Kernel.kt
package com.excelsior.nothing

import com.excelsior.common.QuotedStringTokenizer
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.net.MalformedURLException
import java.net.URL
import java.util.ArrayList
import java.util.HashMap

/**
 * Command execution engine.
 * Translated from Java original by Nikita Lipsky, Excelsior LLC.
 */
object Kernel {

    private val cache = HashMap<String, Boolean>()

    fun isCommand(line: String): Boolean {
        val st = QuotedStringTokenizer(line)
        val cmd = st.nextToken() ?: return false
        cache[cmd]?.let { return it }
        val m = MethodHandle.getMethodHandle(cmd)
        if (m != null) cache[cmd] = true
        return m != null
    }

    fun executeCommand(line: String): Any? {
        cache.clear()
        val st = QuotedStringTokenizer(line)
        val cmd = st.nextToken() ?: return null
        val m = MethodHandle.getMethodHandle(cmd) ?: return null

        val firstArg = st.nextToken()
        val args = ArrayList<String>()

        if (firstArg != null) {
            val selText = AppState.currentSelectionText
            if (firstArg == "^" && selText != null) {
                val selSt = QuotedStringTokenizer(selText)
                while (true) {
                    val tok = selSt.nextToken() ?: break
                    args.add(tok)
                }
            } else {
                args.add(firstArg)
                while (true) {
                    val nextToken = st.nextToken() ?: break
                    args.add(nextToken)
                }
            }
        }

        return try {
            m.invoke(args.toTypedArray())
        } catch (e: Exception) {
            null
        }
    }

    fun addToRegistry(key: String, o: Any) {
        AppState.addToRegistry(key, o)
    }

    fun getFromRegistry(key: String): Any? {
        return AppState.getFromRegistry(key)
    }

    fun checkThreeDots(file: String): String {
        if (file == "...") {
            val dialog = java.awt.FileDialog(null as java.awt.Frame?, "Open")
            dialog.isVisible = true
            return if (dialog.file != null) "${dialog.directory}${dialog.file}" else file
        }
        return file
    }

    fun getInputStream(file: String): InputStream? {
        return try {
            val url = URL(file)
            MethodHandle.addBaseURL(getBaseUrlFrom(url))
            url.openStream()
        } catch (e: MalformedURLException) {
            if (file == "...") return null
            try {
                FileInputStream(file)
            } catch (ioe: IOException) {
                System.err.println("Cannot open file: $file - ${ioe.message}")
                null
            }
        }
    }

    private fun getBaseUrlFrom(url: URL): URL {
        return try {
            URL(url.protocol + "://" + url.host)
        } catch (e: MalformedURLException) {
            throw AssertionError("Bad base URL built")
        }
    }
}
